README.md
# Branch preview

Preview your latest changes. This GitHub Action deploys your branch to a Dokku instance and automatically assigns it a subdomain. This is an easy way to quickly push your latest changes and share your subdomain url for others to preview.

## Usage

Add to your .github workflow the action below.

```yml
name: Branch Preview

on: push

jobs:
  deploy:
    name: Deploy
    runs-on: ubuntu-latest

    steps:
      - name: Checkout code
        uses: actions/checkout@v2

      - name: Preview Branch
        uses: sapiderman/branch-preview-action@v1.0.0
        with:
          dokku_key: ${{ secrets.DOKKU_KEY }}
          host: 111.222.333.444
          port: 22 
          domain: mydomain.com
          github_token: ${{ secrets.GITHUB_TOKEN }}
          github_primary_branch: 'main'
```  

You can refer to other version, or even the main brance like so:

```yml
  uses: sapiderman/branch-preview-action@main
```

For convenience, it may be easier to set the input variables as secrets in GitHub. You can then change the value on the fly.

## Requirements

You will need a dokku instance and the ssh/private key to push your project to. You will also need a domain already pointing to your dokku. Enable virtualhost in dokku setting, so the apps can be mapped to subdomain.

## Input

You need the following requirements. Please store keys and tokens in secret.

| Requirement  | Description                                                                      |  
| :----------- | :--------------------------------------------------------------------------------|  
| host         | Dokku Host address (ip) to push your branch to                                   |  
| port         | Port if your ssh into the host is not default 22                                 |  
| domain       | Name which will be used for your domain.                                         |  
| dokku_key    | Private/SSH Key to your Dokku instance                                           |  
| github_token | Access token. `secrets.GITHUB_TOKEN`, is provided automatically by github workflow|  
| github_primary_branch | (optional) This is your primary git branch, GitHub now defaults to main. set it here |  

## Output

A `subdomain.your-domain.com` is created on your dokku instance. `subdomain` is the current branch other than the primary.

In dokku's view your `subdomain` is just an app. You can check the apps in your dokku by issueing: `dokku apps:list`  

## Troubleshooting

The best way to troubleshoot is if you are able to push from your local to the dokku. All the rights and keys should work from local as from GitHub Actions.

On your local check if this works:

`git push mydokku main:master`

Where `mydokku` is your remote dokku location that you've set using `git remote add mydokku <some address>`.

`main` is just your local git branch as source, it is the default GitHub primary branch. If your local is also master, it would look `master:master` in the command above.

If everything works on the local, it should transfer seamlessly in the GitHub Action. If not, look for the difference.

One note on environment secrets. They do not show up as *** in logs, something to note when debugging.

## Contributing

Find any issues? Feedback? Fork, fix and send the PRs this way!

## Docs

1. [changelog](./CHANGELOG.md)
2. [code of conduct](./code_of_conduct.md)
3. [LICENSE](./LICENSE)

--  
fork. clone. share.


CHANGELOG.md

# CHANGELOG

Highlights the different releases for easy tracking.

## Released
Use this section to tell people about which versions of your project are
currently being supported with security updates.

| Version | Supported                          |
| ------- | ---------------------------------- |
| 1.0.0   | :white_check_mark: Initial release |

## Ongoing

1.x.x - creating some project actions when pushing (stay tuned)

## Reporting a Vulnerability

If you find any issues, or vunerabilities to: xxx@xxx.xx



deploy.sh
#!/bin/bash

echo "Setting up SSH directory"
SSH_PATH="$HOME/.ssh"
mkdir -p "$SSH_PATH"
chmod 700 "$SSH_PATH"

echo "Saving SSH key"
echo "$PRIVATE_KEY" > "$SSH_PATH/deploy_key"
chmod 600 "$SSH_PATH/deploy_key"

echo "Registering key to ssh-agent"
eval $(ssh-agent -s)
ssh-add -k "$SSH_PATH/deploy_key"

echo "Adding hosts key to known_hosts"
touch "$SSH_PATH/known_hosts"
chmod 600 "$SSH_PATH/known_hosts"
ssh-keyscan -t rsa "$HOST" >> "$SSH_PATH/known_hosts"

DEFAULT_BRANCH=${DEFAULT_BRANCH:-'master'}
CURRENT_BRANCH=`git rev-parse --abbrev-ref HEAD`

if [[ $CURRENT_BRANCH == $DEFAULT_BRANCH && -n $DOMAIN_NAME ]]; then
  APP_NAME=$DOMAIN_NAME
else
  APP_NAME=${CURRENT_BRANCH/\//-}
fi

echo "Checking if app exists"
ssh "dokku@$HOST" apps:exists $APP_NAME

if [[ $? != 0 ]]; then
  echo "The app does not exist yet, creating the app"
  ssh "dokku@$HOST" apps:create $APP_NAME
fi

echo "Deploying to host"
git fetch --unshallow
git remote add $APP_NAME "dokku@$HOST:$APP_NAME"
git push -f $APP_NAME "$CURRENT_BRANCH:master"

createProject.js
const { request } = require('@octokit/request');
const { createActionAuth } = require('@octokit/auth-action');

const REPOSITORY = process.env.GITHUB_REPOSITORY;
const BRANCH_NAME = process.argv[2];
const PROJECT_NAME = process.argv[3];

const auth = createActionAuth();

const requestWithAuth = request.defaults({
  request: {
    hook: auth.hook,
  },
});

async function isProjectExists(owner, repo, projectName) {
  const { data: projects } = await requestWithAuth(
    'GET /repos/{owner}/{repo}/projects',
    {
      owner,
      repo,
      mediaType: {
        previews: ['inertia'],
      },
    }
  );

  return projects.some((project) => project.name === projectName);
}

async function createProject(repository) {
  const [owner, repo] = repository.split('/');

  const projectExists = await isProjectExists(owner, repo, PROJECT_NAME);

  if (!projectExists) {
    console.log(`Creating project tracker for ${BRANCH_NAME} branch`);

    const { data: project } = await requestWithAuth(
      'POST /repos/{owner}/{repo}/projects',
      {
        owner,
        repo,
        name: PROJECT_NAME,
        body: `Tracker for ${BRANCH_NAME} branch`,
        mediaType: {
          previews: ['inertia'],
        },
      }
    );

    await createProjectColumn(project.id, 'To do');
    await createProjectColumn(project.id, 'In progress');
    await createProjectColumn(project.id, 'Done');
  }
}

async function createProjectColumn(projectId, columnName) {
  await requestWithAuth('POST /projects/{project_id}/columns', {
    project_id: projectId,
    name: columnName,
    mediaType: {
      previews: ['inertia'],
    },
  });
}

if (REPOSITORY) {
  createProject(REPOSITORY);
}


entrypoint.sh
#!/bin/bash

echo "Starting action"

eval "$(ssh-agent -s)"

echo "Setting up SSH..."
mkdir -p ~/.ssh
chmod 700 ~/.ssh
echo -e "$DOKKU_KEY" > ~/.ssh/id_rsa
chmod 600 ~/.ssh/id_rsa
ssh-keyscan -p "$PORT" "$HOST" >> ~/.ssh/known_hosts

cd "$GITHUB_WORKSPACE" || exit

DEFAULT_BRANCH=${DEFAULT_BRANCH:-$GITHUB_MAIN_BRANCH}
CURRENT_BRANCH=$(git rev-parse --abbrev-ref HEAD)

if [[ "$CURRENT_BRANCH" == "$DEFAULT_BRANCH" && -n "$DOMAIN" ]]; then
  APP_NAME=$DOMAIN
else
  APP_NAME=${CURRENT_BRANCH/\//-}
fi

echo "APP_NAME defined as $APP_NAME"

echo "Checking if app exists"
ssh "dokku@$HOST" -p "$PORT" dokku apps:exists "\$APP_NAME"
RSLT="$?"

if [[ "$RSLT" != 0 ]]; then
  echo "The app does not exist yet, creating the app: $APP_NAME"
  ssh "dokku@$HOST" -p "$PORT" dokku apps:create "\$APP_NAME"
fi

echo "Deploying to host: $HOST"
git fetch --unshallow
git remote add "$APP_NAME" "dokku@$HOST:$PORT/$APP_NAME"
echo "pushing changes to app:$APP_NAME"
git push -f "$APP_NAME" "$CURRENT_BRANCH:master"
echo "done... thank you."


